Assets/MixedRealityToolkit.SDK/Experimental/Features/UX/BoundsControl/Visuals/ProximityEffect/ProximityEffect.cs

```diff
@@ -0,0 +1,353 @@
+
+        #region public methods
+        /// <summary>
+        /// register objects for proximity effect via a <see cref="IProximityEffectObjectProvider"/>
+        /// </summary>
+        public void AddObjects(IProximityEffectObjectProvider provider)
+        {
+            RegisteredObjects registeredObject = new RegisteredObjects() { objectProvider = provider, proximityInfos = new List<ObjectProximityInfo>() };
+            provider.ForEachProximityObject(proximityObject =>
+            {
+                registeredObject.proximityInfos.Add(new ObjectProximityInfo()
+                {
+                    ScaledObject = proximityObject,
+                    ObjectVisualRenderer = proximityObject.gameObject.GetComponentInChildren<Renderer>()
+                });
+            });
+            registeredObjects.Add(registeredObject);
+        }
+
+        /// <summary>
+        /// Clears all registered objects in the proximity effect
+        /// </summary>
+        public void ClearObjects()
+        {
+            if (registeredObjects != null)
+            {
+                registeredObjects.Clear();
+            }
+        }
+
+        /// <summary>
+        /// Resets all objects that had a proximity effect applied. This will reset them to their default size and reset to the base material
+        /// </summary>
+        public void ResetProximityScale()
+        {
+            if (proximityEffectActive == false)
+            {
+                return;
+            }
+
+            foreach (var registeredObject in registeredObjects)
+            {
+                foreach (var item in registeredObject.proximityInfos)
+                {
+                    if (item.ProximityState != ProximityState.FullsizeNoProximity)
+                    {
+                        item.ProximityState = ProximityState.FullsizeNoProximity;
+                       
+                        if (item.ObjectVisualRenderer)
+                        {
+                            item.ObjectVisualRenderer.material = registeredObject.objectProvider.GetBaseMaterial();
+                        }
+
+                        ScaleObject(item.ProximityState, item.ScaledObject, registeredObject.objectProvider.GetObjectSize());
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Updates proximity effect and it's registered objects.
+        /// Highlights and scales objects in proximity according to the pointer distance
+        /// </summary>
+        /// <param name="boundsCenter">gameobject position the proximity effect is attached to</param>
+        /// <param name="boundsExtents">extents of the gameobject the proximity effect is attached to</param>
+        public void Update(Vector3 boundsCenter, Vector3 boundsExtents)
+        {
+            // early out if effect is disabled
+            if (proximityEffectActive == false || !IsAnyRegisteredObjectVisible())
```

Comment:
This area had pretty bad performance right before GA. This update method is probably worth examining again under the profiler since we get to refactor now